javascript/ql/src/semmle/javascript/security/performance/SuperlinearBackTracking.qll

```diff
@@ -388,17 +384,30 @@ private predicate matchesEpsilon(RegExpTerm t) {
   forex(RegExpTerm child | child = t.(RegExpSequence).getAChild() | matchesEpsilon(child))
 }
 
+/**
+ * Gets a message for why `term` can cause polynomial backtracking.
+ */
+string getReasonString(RegExpTerm term, string pump, string prefixMsg, RegExpTerm prev) {
+  polynimalReDoS(term, pump, prefixMsg, prev) and
+  result =
+    "Strings " + prefixMsg + "with many repetitions of '" + pump +
+      "' can start matching anywhere after the start of the preceeding " + prev
+}
+
 /**
  * A term that may cause a regular expression engine to perform a
  * polynomial number of match attempts, relative to the input length.
  */
 class PolynomialBackTrackingTerm extends InfiniteRepetitionQuantifier {
   string reason;
+  string pump;
```

Comment:
Sanity checking for fields vs inline use of predicates: are these four fields functionally dependent? Or can you get some weird combination here?

Reply:
We could get rid of the `reason` field, since there exists a unique `reason` for a `(pump, prefixMsg, prev)` triplet.    
But we need to keep the `(pump, prefixMsg, prev)` triplet, because multiple triplets correspond to the same `reason`. 

So yeah, if we kept just the `reason` field, then we could get weird combinations. 

I think it is practical to the fields as is - that makes it easy to select the shortest possible `reason`.